File: sdks/go/gbz185/transport.go
package gbz185

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

type JSONValue = any
type JSONObject = map[string]any

type JsonTransport interface {
	Request(ctx context.Context, operation string, payload any, out any) error
}

type JsonOperationHandler func(ctx context.Context, payload any) (any, error)

type InProcessJsonTransport struct {
	handlers map[string]JsonOperationHandler
}

func NewInProcessJsonTransport() *InProcessJsonTransport {
	return &InProcessJsonTransport{handlers: map[string]JsonOperationHandler{}}
}

func (t *InProcessJsonTransport) Register(operation string, handler JsonOperationHandler) {
	t.handlers[operation] = handler
}

func (t *InProcessJsonTransport) Request(ctx context.Context, operation string, payload any, out any) error {
	handler, ok := t.handlers[operation]
	if !ok {
		return fmt.Errorf("no JSON transport handler registered for operation: %s", operation)
	}
	result, err := handler(ctx, payload)
	if err != nil {
		return err
	}
	if out == nil {
		return nil
	}
	data, err := json.Marshal(result)
	if err != nil {
		return err
	}
	return json.Unmarshal(data, out)
}

type HttpJsonTransport struct {
	Endpoint string
	Headers  map[string]string
	Client   *http.Client
}

func (t *HttpJsonTransport) Request(ctx context.Context, operation string, payload any, out any) error {
	body, err := json.Marshal(JSONObject{"operation": operation, "payload": payload})
	if err != nil {
		return err
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, t.Endpoint, bytes.NewReader(body))
	if err != nil {
		return err
	}
	req.Header.Set("content-type", "application/json")
	for key, value := range t.Headers {
		req.Header.Set(key, value)
	}
	client := t.Client
	if client == nil {
		client = http.DefaultClient
	}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("HTTP JSON transport failed: %s", resp.Status)
	}
	if out == nil {
		return nil
	}
	return json.NewDecoder(resp.Body).Decode(out)
}

File: sdks/rust/src/transport.rs
use crate::Result;
use serde_json::{json, Value};
use std::cell::RefCell;
use std::collections::HashMap;

pub trait JsonTransport {
    fn request(&self, operation: &str, payload: Value) -> Result<Value>;
}

type Handler = Box<dyn Fn(Value) -> Result<Value>>;

#[derive(Default)]
pub struct InProcessJsonTransport {
    handlers: RefCell<HashMap<String, Handler>>,
}

impl InProcessJsonTransport {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn register<F>(&self, operation: &str, handler: F)
    where
        F: Fn(Value) -> Result<Value> + 'static,
    {
        self.handlers
            .borrow_mut()
            .insert(operation.to_string(), Box::new(handler));
    }
}

impl JsonTransport for InProcessJsonTransport {
    fn request(&self, operation: &str, payload: Value) -> Result<Value> {
        let handlers = self.handlers.borrow();
        let handler = handlers.get(operation).ok_or_else(|| {
            format!("no JSON transport handler registered for operation: {operation}")
        })?;
        handler(payload)
    }
}

pub struct HttpJsonTransport {
    pub endpoint: String,
    pub headers: HashMap<String, String>,
}

impl HttpJsonTransport {
    pub fn new(endpoint: impl Into<String>) -> Self {
        Self {
            endpoint: endpoint.into(),
            headers: HashMap::new(),
        }
    }

    pub fn with_header(mut self, key: impl Into<String>, value: impl Into<String>) -> Self {
        self.headers.insert(key.into(), value.into());
        self
    }
}

impl JsonTransport for HttpJsonTransport {
    fn request(&self, operation: &str, payload: Value) -> Result<Value> {
        let mut req = ureq::post(&self.endpoint).set("content-type", "application/json");
        for (key, value) in &self.headers {
            req = req.set(key, value);
        }
        let response = req.send_json(json!({ "operation": operation, "payload": payload }))?;
        let value: Value = response.into_json()?;
        Ok(value)
    }
}
